node/subsystem/dispatch-gen/src/lib.rs

```diff
@@ -0,0 +1,216 @@
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+use proc_macro2::TokenStream;
+use quote::{quote, ToTokens};
+use std::fmt;
+use syn::{parse2, Error, Fields, FieldsNamed, FieldsUnnamed, Ident, ItemEnum, Path, Result, Type, Variant};
+
+#[proc_macro_attribute]
+pub fn subsystem_dispatch_gen(attr: proc_macro::TokenStream, item: proc_macro::TokenStream) -> proc_macro::TokenStream {
+	let attr = proc_macro2::TokenStream::from(attr);
+	let item = proc_macro2::TokenStream::from(item);
+	proc_macro::TokenStream::from(subsystem_dispatch_gen2(attr, item))
+}
+
+/// An enum variant without base type.
+#[derive(Clone)]
+struct EnumVariantDispatchWithTy {
+	// enum ty name
+	ty: Ident,
+	// variant
+	variant: EnumVariantDispatch,
+}
+
+impl fmt::Debug for EnumVariantDispatchWithTy {
+	fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+		write!(f, "{}::{:?}", self.ty, self.variant)
+	}
+}
+
+impl ToTokens for EnumVariantDispatchWithTy {
+	fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
+		if let Some(inner) = &self.variant.inner {
+			let enum_name = &self.ty;
+			let variant_name = &self.variant.name;
+
+			let ts = quote! {
+				#enum_name::#variant_name(#inner::from(event))
+			};
+
+			tokens.extend(std::iter::once(ts));
+		}
+	}
+}
+
+/// An enum variant without the base type, contains the relevant inner type.
+#[derive(Clone)]
+struct EnumVariantDispatch {
+	/// variant name
+	name: Ident,
+	/// The inner type for which a `From::from` impl is anticipated from the input type.
+	/// No code will be generated for this enum variant if `inner` is `None`.
+	inner: Option<Type>,
+}
+
+impl fmt::Debug for EnumVariantDispatch {
+	fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+		write!(f, "{}(..)", self.name)
+	}
+}
+
+fn prepare_enum_variant(variant: &mut Variant) -> Result<EnumVariantDispatch> {
+	let skip = variant.attrs.iter().find(|attr| attr.path.is_ident("skip")).is_some();
+	variant.attrs = variant.attrs.iter().filter(|attr| !attr.path.is_ident("skip")).cloned().collect::<Vec<_>>();
+
+	let variant = variant.clone();
+	let span = variant.ident.span();
+	let inner = match variant.fields.clone() {
+		// look for one called inner
+		Fields::Named(FieldsNamed { brace_token: _, named }) if !skip => named
+			.iter()
+			.filter(
+				|field| {
+					if let Some(ident) = &field.ident {
+						ident.to_string() == "inner".to_owned()
+					} else {
+						false
+					}
```

Comment:
```suggestion
					if let Some(ident) = &field.ident {
						ident == "inner"
					} else {
						false
					}
```

But can we not also just do `field.ident == Some("inner")`?

Reply:
Doesn't work, since `Option<Ident>` cannot be compared with `Option<&str>`